test/distributed/org/apache/cassandra/distributed/upgrade/ConsistencyTest.java

```diff
@@ -0,0 +1,127 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.upgrade;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.Test;
+
+import org.apache.cassandra.distributed.UpgradeableCluster;
+import org.apache.cassandra.distributed.api.ConsistencyLevel;
+import org.apache.cassandra.distributed.api.Feature;
+import org.apache.cassandra.distributed.api.ICoordinator;
+import org.apache.cassandra.distributed.api.IUpgradeableInstance;
+import org.apache.cassandra.distributed.shared.Versions;
+
+import static org.apache.cassandra.distributed.api.ConsistencyLevel.*;
+import static org.apache.cassandra.distributed.shared.AssertUtils.assertRows;
+import static org.apache.cassandra.distributed.shared.AssertUtils.row;
+
+public class ConsistencyTest extends UpgradeTestBase
+{
+    @Test
+    public void testConsistency() throws Throwable
+    {
+        testConsistency(allUpgrades(3, 1));
+    }
+
+    @Test
+    public void testConsistencyWithNetworkAndGossip() throws Throwable
+    {
+        testConsistency(new TestCase().nodes(3)
+                                      .nodesToUpgrade(1)
+                                      // .upgrade(Versions.Major.v22, Versions.Major.v30)
+                                      // .upgrade(Versions.Major.v22, Versions.Major.v3X)
+                                      // .upgrade(Versions.Major.v30, Versions.Major.v3X)
+                                      .upgrade(Versions.Major.v30, Versions.Major.v4)
+                                      .upgrade(Versions.Major.v3X, Versions.Major.v4)
+                                      .withConfig(cfg -> cfg.with(Feature.NETWORK, Feature.GOSSIP)));
+    }
+
+    private void testConsistency(TestCase testCase) throws Throwable
+    {
+        List<ConsistencyTester> testers = new ArrayList<>();
+        testers.addAll(ConsistencyTester.create(1, ALL, THREE));
+        testers.addAll(ConsistencyTester.create(2, ALL, THREE, QUORUM, LOCAL_QUORUM, TWO));
+        testers.addAll(ConsistencyTester.create(3, ALL, THREE, QUORUM, LOCAL_QUORUM, TWO, ONE, LOCAL_ONE));
+
+        testCase.setup(cluster -> {
+            ConsistencyTester.createTable(cluster);
+            for (ConsistencyTester tester : testers)
+                tester.writeRows(cluster);
+        }).runAfterNodeAndClusterUpgrade(cluster -> {
```

Comment:
@adelapena If we're only upgrading one node, wouldn't using `runAfterNodeAndClusterUpgrade()` just mean we run the supplied lambda twice with the cluster in the same exact state? I'm wondering if we actually need `runAfterNodeAndClusterUpgrade()`. If we upgraded all nodes, `runAfterNodeUpgrade()` would already ensure we run when the cluster is fully upgraded, but if we don't update all the nodes, `runAfterClusterUpgrade()` still runs in a mixed cluster.

Reply:
One thing we could do in the post-cluster-upgrade section is progressively stop nodes depending on the read CL like `TestAvailability` does. The only problem with that seems to be the way the CLs are grouped into a `ConsistencyTester` right now. I suppose those could be sliced the other way though (the ALL/THREE tests, then the QUORUM tests, etc.) and node stops performed between them...

Reply:
I have added a new `MixedModeAvailabilityTest` to verify the behaviour with down nodes, messing with internode messaging. I'm limiting the number of tested consistency levels because, as we said in CASSANDRA-16181, the details of each CL are not the main purpose of the test.